Tim's onboarding challenge

caiicar293 opened the pull request:
# Purpose
Explain the purpose of the PR here, including references to any existing Notion tasks.

# New Changes
- Explain new changes

# Testing
- Explain tests that you ran to verify code functionality.

# Outstanding Changes
- If there are non-critical changes (i.e. additional features) that can be made to this feature in the future, indicate them here.

caiicar293 committed:
Update controller.h

caiicar293 committed:
Level2
# Purpose
Explain the purpose of the PR here, including references to any existing Notion tasks.

# New Changes
- Explain new changes

# Testing
- Explain tests that you ran to verify code functionality.

# Outstanding Changes
- If there are non-critical changes (i.e. additional features) that can be made to this feature in the future, indicate them here.

caiicar293 committed:
Merge branch 'level2' of https://github.com/UWOrbital/OBC-Firmware-Onboarding into level2

caiicar293 committed:
Merge branch 'level2' of https://github.com/caiicar293/OBC-Firmware-Onboarding into level2

dgobalak reviewed (changes requested):
Looks like there are some build errors you need to fix.

kiransuren commented:
Looks like you have some git conflict marker artifacts in your code? If you need some help fixing them please reach out to Daniel, Chad, George, Nathan or I

caiicar293 committed:
Tim onboarding challenge attempt 2
I managed to fix the build errors I believe

# New Changes
- Explain new changes

# Testing
- Explain tests that you ran to verify code functionality.

# Outstanding Changes
- If there are non-critical changes (i.e. additional features) that can be made to this feature in the future, indicate them here.

caiicar293 commented:
> Looks like you have some git conflict marker artifacts in your code? If you need some help fixing them please reach out to Daniel, Chad, George, Nathan or I

dgobalak commented on drivers/source/serial_io.c:
```diff
@@ -19,7 +19,7 @@
 /* USER CODE BEGIN */
 // Add a static assertion to ensure that UART_PRINT_REG is defined as either scilinREG or sciREG
-
+STATIC_ASSERT(UART_PRINT_REG == scilinREG, "UART_PRINT_REG does not equal scilinREG");
```
`UART_PRINT_REG` can be _either_ `scilinREG` or `sciREG`.

dgobalak commented on drivers/source/serial_io.c:
```diff
@@ -64,9 +66,28 @@ obc_error_code_t sciPrintText(unsigned char *text, uint32_t length) {
     /* USER CODE BEGIN */
     // Print text to the serial port using sciSendBytes. Use the mutex to protect the SCI module.
-
+    ASSERT(sciMutex != NULL);
```
No need for these assertions. That's what the above if statement is for.

dgobalak commented on drivers/source/serial_io.c:
```diff
@@ -64,9 +66,28 @@ obc_error_code_t sciPrintText(unsigned char *text, uint32_t length) {
+    ASSERT(sciMutex != NULL);
+        
+
+    if (sciMutex != NULL) {
```
This check is done earlier.

dgobalak commented on drivers/source/serial_io.c:
```diff
@@ -64,9 +66,28 @@ obc_error_code_t sciPrintText(unsigned char *text, uint32_t length) {
+    }
+    ASSERT(sciLinMutex != NULL);
+        
+    if (sciLinMutex != NULL) {
```
This check is done earlier.

dgobalak commented on drivers/source/serial_io.c:
```diff
@@ -64,9 +66,28 @@ obc_error_code_t sciPrintText(unsigned char *text, uint32_t length) {
+        if (xSemaphoreTake(sciMutex, UART_MUTEX_BLOCK_TIME) == pdTRUE) {
+            sciSendBytes(text, length);
+            xSemaphoreGive(sciMutex);
+            return 1;
```
You should be returning the relevant error code, not just 0 or 1. The defined error codes can be found in `obc_errors.h`. Notice how the function return type is `obc_error_code_t`

dgobalak commented on drivers/source/serial_io.c:
```diff
@@ -64,9 +66,28 @@ obc_error_code_t sciPrintText(unsigned char *text, uint32_t length) {
+    if (sciMutex != NULL) {
+        if (xSemaphoreTake(sciMutex, UART_MUTEX_BLOCK_TIME) == pdTRUE) {
+            sciSendBytes(text, length);
+            xSemaphoreGive(sciMutex);
```
You should be printing to the serial port associated with `UART_PRINT_REG`. This means that the mutex you need to acquire depends on what `UART_PRINT_REG` is defined as. Here, you're not checking `UART_PRINT_REG`. However, you also don't need to check it. That's done for you on the `SemaphoreHandle_t mutex = (UART_PRINT_REG == sciREG) ? sciMutex : sciLinMutex;` line. Just use that mutex instead of `sciMutex` or `scilinMutex`

dgobalak commented on drivers/source/serial_io.c:
```diff
@@ -64,9 +66,28 @@ obc_error_code_t sciPrintText(unsigned char *text, uint32_t length) {
+            return 1;
+        }
+    }
+    return 0;    
```
Return the relevant error code

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +47,62 @@ static void lightServiceTask(void * pvParameters);
     /* USER CODE BEGIN */
-    // Create the task and queue here. Return error code if task/queue was not created successfully.
+    BaseType_t xReturned = pdFAIL;
+        if (lightServiceHandle == NULL) {
```
Looks like your indentation is off

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +47,62 @@ static void lightServiceTask(void * pvParameters);
+        }
+
+        if (lightserviceQueue == NULL) {
+                lightserviceQueue = xQueueCreate(LIGHT_SERVICE_SIZE, LIGHT_QUEUE_SIZE);
```
Indentation is off here as well

dgobalak replied:
We use the statically-allocated version of tasks, queues, etc. In this case, use `xQueueCreateStatic` instead.

dgobalak commented on tempCodeRunnerFile.c:
```diff
@@ -0,0 +1,10 @@
+
```
Not sure what this file is. You can remove it

dgobalak commented on onboarding/source/controller.c:
```diff
@@ -26,7 +27,9 @@
 /* USER CODE BEGIN */
 // define config for the light timer
-
+#define LIGHT_CONTROLLER_NAME         "light timer" 
```
These macro names are misleading. The "light controller" is a task, whereas this is a timer to send an event to that task.

dgobalak commented on onboarding/source/controller.c:
```diff
@@ -85,9 +89,24 @@ obc_error_code_t initController(void) {
+    }
+                                        
+    if (lightTimerHandle == NULL) 
+            return OBC_ERR_CODE_TIMER_CREATION_FAILED;
```
It looks like the indentation is off here too. There's probably some setting in your IDE that you should change. Indentation should be 4 spaces

dgobalak commented on onboarding/source/controller.c:
```diff
@@ -85,9 +89,24 @@ obc_error_code_t initController(void) {
-    /* USER CODE END */
+        /* USER CODE END */
+
+            return 1;
```
Return the relevant error code

dgobalak commented on onboarding/source/controller.c:
```diff
@@ -97,16 +116,24 @@ static void controllerTask(void * pvParameters) {
         /* USER CODE BEGIN */
-        // Deal with error when initializing light service task and/or queue
+        char str[] = "Error: Light service was not properly initialized";
+        sciPrintText((unsigned char *) str, strlen(str));
```
Note that there's a `sciPrintf` function that behaves very similarly to the stdio `printf` function. It's cleaner to use that instead.

dgobalak commented on onboarding/source/controller.c:
```diff
@@ -97,16 +116,24 @@ static void controllerTask(void * pvParameters) {
+ 
+        Light service task and queue created successfully */
         BaseType_t xReturned; 
         xReturned = xTimerStart(ledTimerHandle, 0);
```
Check if this was started successfully.

dgobalak commented on onboarding/source/controller.c:
```diff
@@ -97,16 +116,24 @@ static void controllerTask(void * pvParameters) {
+
+        if (xReturned == pdFAIL) {
+             char str[] = "Error: Light Timer was not properly initialized";
+             sciPrintText((unsigned char *) str, strlen(str));
```
Note that there's a `sciPrintf` function that behaves very similarly to the stdio `printf` function. It's cleaner to use that instead.

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -2,10 +2,19 @@
+#include <FreeRTOS.h>
+#include <os_task.h>
+#include <os_queue.h>
+#include <adc.h>
```
These are already included at the top of the file

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +47,62 @@ static void lightServiceTask(void * pvParameters);
-    // Create the task and queue here. Return error code if task/queue was not created successfully.
+    BaseType_t xReturned = pdFAIL;
+        if (lightServiceHandle == NULL) {
+            xReturned = xTaskCreate(lightServiceTask,
```
We use the statically-allocated version of tasks, queues, etc. In this case, use `xTaskCreateStatic` instead.

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +47,62 @@ static void lightServiceTask(void * pvParameters);
+
+                char str[32];
+                    snprintf(str, 32, "The ambient light is %d", adcData.value);
+                    sciPrintText((unsigned char *) str, strlen(str));
```
Note that there's a `sciPrintf` function that behaves very similarly to the stdio `printf` function. It's cleaner to use that instead.

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +47,62 @@ static void lightServiceTask(void * pvParameters);
+    if(lightserviceQueue != NULL){
+        xReturned = xQueueSend(lightserviceQueue, event, portMAX_DELAY);
+    }
+    return xReturned;
```
Return the relevant error code

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +47,62 @@ static void lightServiceTask(void * pvParameters);
+
+
+ 	while (1) {
+        if (xQueueReceive(lightserviceQueue, &eventReceived, portMAX_DELAY) == pdPASS) {
```
Indentation is off

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +47,62 @@ static void lightServiceTask(void * pvParameters);
+                     return OBC_ERR_CODE_QUEUE_FULL;
+                }
+        }
+    return 1;                            
```
Return the relevant error code

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +47,62 @@ static void lightServiceTask(void * pvParameters);
+                adcGetData(adcREG1, adcGROUP1, &adcData);
+
+                char str[32];
+                    snprintf(str, 32, "The ambient light is %d", adcData.value);
```
`adcData.value` is a `uint16_t` if I remember correctly. Use the `%u` format specifier instead

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +47,62 @@ static void lightServiceTask(void * pvParameters);
+        if (xQueueReceive(lightserviceQueue, &eventReceived, portMAX_DELAY) == pdPASS) {
+            adcData_t adcData;
+            if (eventReceived == MEASURE_LIGHT) {
+                adcStartConversion(adcREG1, adcGROUP1);
```
It'd be better to create a separate helper function to get the ADC value. It's important to keep your code modular. It makes it much easier to test and minimizes repetitive code.

dgobalak reviewed (changes requested):
Great start! Your code _should_ work for the most part (just a few bugs). Most of my comments are related to writing code specifically for embedded systems. I also commented on some things related to making your code more readable.

caiicar293 commented:
Suggested edit:
```diff
diff --git a/drivers/source/serial_io.c b/drivers/source/serial_io.c
index 83d8b13..9c89e29 100644
--- a/drivers/source/serial_io.c
+++ b/drivers/source/serial_io.c
@@ -19,7 +19,7 @@
 
 /* USER CODE BEGIN */
 // Add a static assertion to ensure that UART_PRINT_REG is defined as either scilinREG or sciREG
-STATIC_ASSERT(UART_PRINT_REG == scilinREG, "UART_PRINT_REG does not equal scilinREG");
+STATIC_ASSERT(UART_PRINT_REG == scilinREG||UART_PRINT_REG == sciREG, "UART_PRINT_REG does not equal scilinREG");
 /* USER CODE END */
 STATIC_ASSERT(MAX_PRINTF_SIZE > 0, "MAX_PRINTF_SIZE must be greater than 0");
 
@@ -66,26 +66,20 @@ obc_error_code_t sciPrintText(unsigned char *text, uint32_t length) {
 
     /* USER CODE BEGIN */
     // Print text to the serial port using sciSendBytes. Use the mutex to protect the SCI module.
-    ASSERT(sciMutex != NULL);
+    
         
 
-    if (sciMutex != NULL) {
-        if (xSemaphoreTake(sciMutex, UART_MUTEX_BLOCK_TIME) == pdTRUE) {
-            sciSendBytes(text, length);
-            xSemaphoreGive(sciMutex);
-            return 1;
-        }
-    }
-    ASSERT(sciLinMutex != NULL);
-        
-    if (sciLinMutex != NULL) {
-        if (xSemaphoreTake(sciLinMutex, UART_MUTEX_BLOCK_TIME) == pdTRUE) {
-            sciSendBytes(text, length);
-            xSemaphoreGive(sciLinMutex);
-            return 1;
-        }
+    if (xSemaphoreTake(mutex, UART_MUTEX_BLOCK_TIME) == pdTRUE) {
+        sciSendBytes(text, length);
+        xSemaphoreGive(mutex);
+    } else {
+        return OBC_ERR_CODE_MUTEX_CREATION_FAILED;
     }
-    return 0;    
+    
+    
+
+
+    return OBC_ERR_CODE_SUCCESS;    
     /* USER CODE END */
     }
 
diff --git a/onboarding/include/amb_light_service.h b/onboarding/include/amb_light_service.h
index 5bbfa91..908c287 100644
--- a/onboarding/include/amb_light_service.h
+++ b/onboarding/include/amb_light_service.h
@@ -15,6 +15,15 @@ typedef enum {
  */
 obc_error_code_t initLightService(void);
 
+/**
+ * @brief get data from ADC
+ *
+ * @param void
+ * @note use in lightServiceTask(void *pvParameters)
+ * @return uint32, return data value from sensor
+ */
+uint32_t getLightSensorData(void);
+
 /**
  * @brief Send an event to the light service queue.
  * 
diff --git a/onboarding/source/amb_light_service.c b/onboarding/source/amb_light_service.c
index 896c946..a38f124 100644
--- a/onboarding/source/amb_light_service.c
+++ b/onboarding/source/amb_light_service.c
@@ -12,8 +12,7 @@
 #include <FreeRTOS.h>
 #include <os_task.h>
 #include <os_queue.h>
-#include <adc.h>
-#include <sci.h>
+
 
 // Include any additional headers here
 
@@ -26,9 +25,9 @@
 
 /* USER CODE BEGIN */
 // Define light service queue config here
-#define LIGHT_SERVICE_SIZE 10
+#define LIGHT_QUEUE_LENGTH 10
 #define LIGHT_QUEUE_SIZE sizeof(light_event_t)
-#define TEXT_SIZE 8
+#define TEXT_SIZE sizeof( uint64_t )
 /* USER CODE END */
 
 /* USER CODE BEGIN */
@@ -36,7 +35,13 @@
 
 
 static TaskHandle_t lightServiceHandle = NULL;
+static StaticTask_t lightServiceBuffer;
+static StackType_t lightServiceStack[LIGHT_SERVICE_STACK_SIZE];
+
 static QueueHandle_t lightserviceQueue;
+static StaticQueue_t xlightStaticQueue;
+uint8_t uclightQueueStorageArea[LIGHT_QUEUE_LENGTH * TEXT_SIZE];
+
 /* USER CODE END */
 
 /**
@@ -47,29 +52,50 @@ static void lightServiceTask(void * pvParameters);
 
 obc_error_code_t initLightService(void) {
     /* USER CODE BEGIN */
-    BaseType_t xReturned = pdFAIL;
-        if (lightServiceHandle == NULL) {
-            xReturned = xTaskCreate(lightServiceTask,
-                                    LIGHT_SERVICE_NAME,
-                                    LIGHT_SERVICE_STACK_SIZE,
-                                    NULL,
-                                    LIGHT_SERVICE_PRIORITY,
-                                    &lightServiceHandle);
-                                     }
-        if (xReturned == pdFAIL) {
-             return OBC_ERR_CODE_TASK_CREATION_FAILED;
-        
-        }
+    if (lightServiceHandle == NULL) {
+        lightServiceHandle = xTaskCreateStatic(lightServiceTask,
+                                LIGHT_SERVICE_NAME,
+                                LIGHT_SERVICE_STACK_SIZE,
+                                NULL,
+                                LIGHT_SERVICE_PRIORITY,
+                                lightServiceStack,
+                                &lightServiceBuffer);
+                                    }
+    if (lightServiceHandle == NULL)
+    {
+        return OBC_ERR_CODE_TASK_CREATION_FAILED;
+    }
 
-        if (lightserviceQueue == NULL) {
-                lightserviceQueue = xQueueCreate(LIGHT_SERVICE_SIZE, LIGHT_QUEUE_SIZE);
+    if (lightserviceQueue == NULL) {
+        lightserviceQueue = xQueueCreateStatic(LIGHT_QUEUE_LENGTH, LIGHT_QUEUE_SIZE,uclightQueueStorageArea, &xlightStaticQueue);
 
-                if (lightserviceQueue == NULL) {
-                     return OBC_ERR_CODE_QUEUE_FULL;
-                }
-        }
-    return 1;                            
-    /* USER CODE END */
+    if (lightserviceQueue == NULL) {
+        return OBC_ERR_CODE_INVALID_ARG;
+    }
+    }
+return OBC_ERR_CODE_SUCCESS;                           
+/* USER CODE END */
+}
+
+uint32_t getLightSensorData(void)
+{
+	adcData_t adcData;
+	
+
+ 	/** - Start Group1 ADC Conversion 
+ 	*     Select Channel 6 - Light Sensor for Conversion
+ 	*/
+	adcStartConversion(adcREG1, adcGROUP1);
+
+ 	/** - Wait for ADC Group1 conversion to complete */
+ 	while(!adcIsConversionComplete(adcREG1, adcGROUP1)); 
+
+	/** - Read the conversion result
+	*     The data contains the Light sensor data
+    */
+	adcGetData(adcREG1, adcGROUP1, &adcData);
+	
+	return adcData.value;
 }
 
 static void lightServiceTask(void * pvParameters) {
@@ -77,32 +103,32 @@ static void lightServiceTask(void * pvParameters) {
     // Wait for MEASURE_LIGHT event in the queue and then print the ambient light value to the serial port.
     light_event_t eventReceived;
 
-
- 	while (1) {
+    while (1) {
         if (xQueueReceive(lightserviceQueue, &eventReceived, portMAX_DELAY) == pdPASS) {
-            adcData_t adcData;
+            
             if (eventReceived == MEASURE_LIGHT) {
-                adcStartConversion(adcREG1, adcGROUP1);
-                while (!adcIsConversionComplete(adcREG1, adcGROUP1));
-                adcGetData(adcREG1, adcGROUP1, &adcData);
+                
 
                 char str[32];
-                    snprintf(str, 32, "The ambient light is %d", adcData.value);
-                    sciPrintText((unsigned char *) str, strlen(str));
-                }
+                uint16_t lightdata = getLightSensorData();
+                    snprintf(str, 32, "The ambient light is %u", lightdata);
+                    /** - Transmit the Conversion data to PC using SCI*/
+                    sciPrintf("%u\n", str);
+             }
 
                 
-            }
+          }
     }
     /* USER CODE END */
-}
+ }
 
 obc_error_code_t sendToLightServiceQueue(light_event_t *event) {
     /* USER CODE BEGIN */
-    BaseType_t xReturned = pdFAIL;
     if(lightserviceQueue != NULL){
-        xReturned = xQueueSend(lightserviceQueue, event, portMAX_DELAY);
+        if(xQueueSend(lightserviceQueue, event, portMAX_DELAY != pdTRUE)){
+            return OBC_ERR_CODE_QUEUE_FULL;
+        }
     }
-    return xReturned;
+    return OBC_ERR_CODE_SUCCESS;
     /* USER CODE END */
 }
diff --git a/onboarding/source/controller.c b/onboarding/source/controller.c
index aac7564..5e7275e 100644
--- a/onboarding/source/controller.c
+++ b/onboarding/source/controller.c
@@ -27,9 +27,9 @@
 
 /* USER CODE BEGIN */
 // define config for the light timer
-#define LIGHT_CONTROLLER_NAME         "light timer" 
-#define LIGHT_CONTROLLER_PERIOD        pdMS_TO_TICKS(1000)
-#define LIGHT_CONTROLLER_AUTORELOAD    pdTRUE 
+#define LIGHT_TIMER_NAME         "light timer" 
+#define LIGHT_TIMER_PERIOD        pdMS_TO_TICKS(1000)
+#define LIGHT_TIMER_AUTORELOAD    pdTRUE 
 /* USER CODE END */
 
 /* Declare handlers and buffers for tasks and timers */
@@ -91,22 +91,22 @@ obc_error_code_t initController(void) {
     /* USER CODE BEGIN */
     if (lightTimerHandle == NULL) {
         // Create light timer
-        lightTimerHandle = xTimerCreateStatic(LIGHT_CONTROLLER_NAME,
-                                        LIGHT_CONTROLLER_PERIOD, 
-                                        LIGHT_CONTROLLER_AUTORELOAD,
+        lightTimerHandle = xTimerCreateStatic(LIGHT_TIMER_NAME,
+                                        LIGHT_TIMER_PERIOD, 
+                                        LIGHT_TIMER_AUTORELOAD,
                                         (void *) 0,
                                         lightTimerCallback,
                                         &lightTimerBuffer);
     }
                                         
     if (lightTimerHandle == NULL) 
-            return OBC_ERR_CODE_TIMER_CREATION_FAILED;
+        return OBC_ERR_CODE_TIMER_CREATION_FAILED;
             
         
 
         /* USER CODE END */
 
-            return 1;
+    return OBC_ERR_CODE_SUCCESS;
 }
 
 static void controllerTask(void * pvParameters) {
@@ -117,7 +117,7 @@ static void controllerTask(void * pvParameters) {
     if (lightServiceStatus != OBC_ERR_CODE_SUCCESS) {
         /* USER CODE BEGIN */
         char str[] = "Error: Light service was not properly initialized";
-        sciPrintText((unsigned char *) str, strlen(str));
+        sciPrintf("%s\n",str);
 
     } else { 
         /* USER CODE END 
@@ -125,13 +125,18 @@ static void controllerTask(void * pvParameters) {
         Light service task and queue created successfully */
         BaseType_t xReturned; 
         xReturned = xTimerStart(ledTimerHandle, 0);
+
+        if (xReturned == pdFAIL) {
+             char str[] = "Error: LED Timer was not properly initialized";
+             sciPrintf("%s\n",str);
+        }
         
         /* USER CODE BEGIN */
         xReturned = xTimerStart(lightTimerHandle, 0);
 
         if (xReturned == pdFAIL) {
              char str[] = "Error: Light Timer was not properly initialized";
-             sciPrintText((unsigned char *) str, strlen(str));
+             sciPrintf("%s\n",str);
         }
             
 
diff --git a/tempCodeRunnerFile.c b/tempCodeRunnerFile.c
deleted file mode 100644
index dfa038f..0000000
--- a/tempCodeRunnerFile.c
+++ /dev/null
@@ -1,10 +0,0 @@
-
-#include <stdio.h>
-
-int main () {
-
-    int a = 23;
-    int *var; 
-    var = &a;
-    printf("%x", a);
-}
\ No newline at end of file

```

caiicar293 commented:
Suggested edit:
```diff
diff --git a/drivers/source/serial_io.c b/drivers/source/serial_io.c
index 83d8b13..9c89e29 100644
--- a/drivers/source/serial_io.c
+++ b/drivers/source/serial_io.c
@@ -19,7 +19,7 @@
 
 /* USER CODE BEGIN */
 // Add a static assertion to ensure that UART_PRINT_REG is defined as either scilinREG or sciREG
-STATIC_ASSERT(UART_PRINT_REG == scilinREG, "UART_PRINT_REG does not equal scilinREG");
+STATIC_ASSERT(UART_PRINT_REG == scilinREG||UART_PRINT_REG == sciREG, "UART_PRINT_REG does not equal scilinREG");
 /* USER CODE END */
 STATIC_ASSERT(MAX_PRINTF_SIZE > 0, "MAX_PRINTF_SIZE must be greater than 0");
 
@@ -66,26 +66,20 @@ obc_error_code_t sciPrintText(unsigned char *text, uint32_t length) {
 
     /* USER CODE BEGIN */
     // Print text to the serial port using sciSendBytes. Use the mutex to protect the SCI module.
-    ASSERT(sciMutex != NULL);
+    
         
 
-    if (sciMutex != NULL) {
-        if (xSemaphoreTake(sciMutex, UART_MUTEX_BLOCK_TIME) == pdTRUE) {
-            sciSendBytes(text, length);
-            xSemaphoreGive(sciMutex);
-            return 1;
-        }
-    }
-    ASSERT(sciLinMutex != NULL);
-        
-    if (sciLinMutex != NULL) {
-        if (xSemaphoreTake(sciLinMutex, UART_MUTEX_BLOCK_TIME) == pdTRUE) {
-            sciSendBytes(text, length);
-            xSemaphoreGive(sciLinMutex);
-            return 1;
-        }
+    if (xSemaphoreTake(mutex, UART_MUTEX_BLOCK_TIME) == pdTRUE) {
+        sciSendBytes(text, length);
+        xSemaphoreGive(mutex);
+    } else {
+        return OBC_ERR_CODE_MUTEX_CREATION_FAILED;
     }
-    return 0;    
+    
+    
+
+
+    return OBC_ERR_CODE_SUCCESS;    
     /* USER CODE END */
     }
 
diff --git a/onboarding/include/amb_light_service.h b/onboarding/include/amb_light_service.h
index 5bbfa91..908c287 100644
--- a/onboarding/include/amb_light_service.h
+++ b/onboarding/include/amb_light_service.h
@@ -15,6 +15,15 @@ typedef enum {
  */
 obc_error_code_t initLightService(void);
 
+/**
+ * @brief get data from ADC
+ *
+ * @param void
+ * @note use in lightServiceTask(void *pvParameters)
+ * @return uint32, return data value from sensor
+ */
+uint32_t getLightSensorData(void);
+
 /**
  * @brief Send an event to the light service queue.
  * 
diff --git a/onboarding/source/amb_light_service.c b/onboarding/source/amb_light_service.c
index 896c946..a38f124 100644
--- a/onboarding/source/amb_light_service.c
+++ b/onboarding/source/amb_light_service.c
@@ -12,8 +12,7 @@
 #include <FreeRTOS.h>
 #include <os_task.h>
 #include <os_queue.h>
-#include <adc.h>
-#include <sci.h>
+
 
 // Include any additional headers here
 
@@ -26,9 +25,9 @@
 
 /* USER CODE BEGIN */
 // Define light service queue config here
-#define LIGHT_SERVICE_SIZE 10
+#define LIGHT_QUEUE_LENGTH 10
 #define LIGHT_QUEUE_SIZE sizeof(light_event_t)
-#define TEXT_SIZE 8
+#define TEXT_SIZE sizeof( uint64_t )
 /* USER CODE END */
 
 /* USER CODE BEGIN */
@@ -36,7 +35,13 @@
 
 
 static TaskHandle_t lightServiceHandle = NULL;
+static StaticTask_t lightServiceBuffer;
+static StackType_t lightServiceStack[LIGHT_SERVICE_STACK_SIZE];
+
 static QueueHandle_t lightserviceQueue;
+static StaticQueue_t xlightStaticQueue;
+uint8_t uclightQueueStorageArea[LIGHT_QUEUE_LENGTH * TEXT_SIZE];
+
 /* USER CODE END */
 
 /**
@@ -47,29 +52,50 @@ static void lightServiceTask(void * pvParameters);
 
 obc_error_code_t initLightService(void) {
     /* USER CODE BEGIN */
-    BaseType_t xReturned = pdFAIL;
-        if (lightServiceHandle == NULL) {
-            xReturned = xTaskCreate(lightServiceTask,
-                                    LIGHT_SERVICE_NAME,
-                                    LIGHT_SERVICE_STACK_SIZE,
-                                    NULL,
-                                    LIGHT_SERVICE_PRIORITY,
-                                    &lightServiceHandle);
-                                     }
-        if (xReturned == pdFAIL) {
-             return OBC_ERR_CODE_TASK_CREATION_FAILED;
-        
-        }
+    if (lightServiceHandle == NULL) {
+        lightServiceHandle = xTaskCreateStatic(lightServiceTask,
+                                LIGHT_SERVICE_NAME,
+                                LIGHT_SERVICE_STACK_SIZE,
+                                NULL,
+                                LIGHT_SERVICE_PRIORITY,
+                                lightServiceStack,
+                                &lightServiceBuffer);
+                                    }
+    if (lightServiceHandle == NULL)
+    {
+        return OBC_ERR_CODE_TASK_CREATION_FAILED;
+    }
 
-        if (lightserviceQueue == NULL) {
-                lightserviceQueue = xQueueCreate(LIGHT_SERVICE_SIZE, LIGHT_QUEUE_SIZE);
+    if (lightserviceQueue == NULL) {
+        lightserviceQueue = xQueueCreateStatic(LIGHT_QUEUE_LENGTH, LIGHT_QUEUE_SIZE,uclightQueueStorageArea, &xlightStaticQueue);
 
-                if (lightserviceQueue == NULL) {
-                     return OBC_ERR_CODE_QUEUE_FULL;
-                }
-        }
-    return 1;                            
-    /* USER CODE END */
+    if (lightserviceQueue == NULL) {
+        return OBC_ERR_CODE_INVALID_ARG;
+    }
+    }
+return OBC_ERR_CODE_SUCCESS;                           
+/* USER CODE END */
+}
+
+uint32_t getLightSensorData(void)
+{
+	adcData_t adcData;
+	
+
+ 	/** - Start Group1 ADC Conversion 
+ 	*     Select Channel 6 - Light Sensor for Conversion
+ 	*/
+	adcStartConversion(adcREG1, adcGROUP1);
+
+ 	/** - Wait for ADC Group1 conversion to complete */
+ 	while(!adcIsConversionComplete(adcREG1, adcGROUP1)); 
+
+	/** - Read the conversion result
+	*     The data contains the Light sensor data
+    */
+	adcGetData(adcREG1, adcGROUP1, &adcData);
+	
+	return adcData.value;
 }
 
 static void lightServiceTask(void * pvParameters) {
@@ -77,32 +103,32 @@ static void lightServiceTask(void * pvParameters) {
     // Wait for MEASURE_LIGHT event in the queue and then print the ambient light value to the serial port.
     light_event_t eventReceived;
 
-
- 	while (1) {
+    while (1) {
         if (xQueueReceive(lightserviceQueue, &eventReceived, portMAX_DELAY) == pdPASS) {
-            adcData_t adcData;
+            
             if (eventReceived == MEASURE_LIGHT) {
-                adcStartConversion(adcREG1, adcGROUP1);
-                while (!adcIsConversionComplete(adcREG1, adcGROUP1));
-                adcGetData(adcREG1, adcGROUP1, &adcData);
+                
 
                 char str[32];
-                    snprintf(str, 32, "The ambient light is %d", adcData.value);
-                    sciPrintText((unsigned char *) str, strlen(str));
-                }
+                uint16_t lightdata = getLightSensorData();
+                    snprintf(str, 32, "The ambient light is %u", lightdata);
+                    /** - Transmit the Conversion data to PC using SCI*/
+                    sciPrintf("%u\n", str);
+             }
 
                 
-            }
+          }
     }
     /* USER CODE END */
-}
+ }
 
 obc_error_code_t sendToLightServiceQueue(light_event_t *event) {
     /* USER CODE BEGIN */
-    BaseType_t xReturned = pdFAIL;
     if(lightserviceQueue != NULL){
-        xReturned = xQueueSend(lightserviceQueue, event, portMAX_DELAY);
+        if(xQueueSend(lightserviceQueue, event, portMAX_DELAY != pdTRUE)){
+            return OBC_ERR_CODE_QUEUE_FULL;
+        }
     }
-    return xReturned;
+    return OBC_ERR_CODE_SUCCESS;
     /* USER CODE END */
 }
diff --git a/onboarding/source/controller.c b/onboarding/source/controller.c
index aac7564..5e7275e 100644
--- a/onboarding/source/controller.c
+++ b/onboarding/source/controller.c
@@ -27,9 +27,9 @@
 
 /* USER CODE BEGIN */
 // define config for the light timer
-#define LIGHT_CONTROLLER_NAME         "light timer" 
-#define LIGHT_CONTROLLER_PERIOD        pdMS_TO_TICKS(1000)
-#define LIGHT_CONTROLLER_AUTORELOAD    pdTRUE 
+#define LIGHT_TIMER_NAME         "light timer" 
+#define LIGHT_TIMER_PERIOD        pdMS_TO_TICKS(1000)
+#define LIGHT_TIMER_AUTORELOAD    pdTRUE 
 /* USER CODE END */
 
 /* Declare handlers and buffers for tasks and timers */
@@ -91,22 +91,22 @@ obc_error_code_t initController(void) {
     /* USER CODE BEGIN */
     if (lightTimerHandle == NULL) {
         // Create light timer
-        lightTimerHandle = xTimerCreateStatic(LIGHT_CONTROLLER_NAME,
-                                        LIGHT_CONTROLLER_PERIOD, 
-                                        LIGHT_CONTROLLER_AUTORELOAD,
+        lightTimerHandle = xTimerCreateStatic(LIGHT_TIMER_NAME,
+                                        LIGHT_TIMER_PERIOD, 
+                                        LIGHT_TIMER_AUTORELOAD,
                                         (void *) 0,
                                         lightTimerCallback,
                                         &lightTimerBuffer);
     }
                                         
     if (lightTimerHandle == NULL) 
-            return OBC_ERR_CODE_TIMER_CREATION_FAILED;
+        return OBC_ERR_CODE_TIMER_CREATION_FAILED;
             
         
 
         /* USER CODE END */
 
-            return 1;
+    return OBC_ERR_CODE_SUCCESS;
 }
 
 static void controllerTask(void * pvParameters) {
@@ -117,7 +117,7 @@ static void controllerTask(void * pvParameters) {
     if (lightServiceStatus != OBC_ERR_CODE_SUCCESS) {
         /* USER CODE BEGIN */
         char str[] = "Error: Light service was not properly initialized";
-        sciPrintText((unsigned char *) str, strlen(str));
+        sciPrintf("%s\n",str);
 
     } else { 
         /* USER CODE END 
@@ -125,13 +125,18 @@ static void controllerTask(void * pvParameters) {
         Light service task and queue created successfully */
         BaseType_t xReturned; 
         xReturned = xTimerStart(ledTimerHandle, 0);
+
+        if (xReturned == pdFAIL) {
+             char str[] = "Error: LED Timer was not properly initialized";
+             sciPrintf("%s\n",str);
+        }
         
         /* USER CODE BEGIN */
         xReturned = xTimerStart(lightTimerHandle, 0);
 
         if (xReturned == pdFAIL) {
              char str[] = "Error: Light Timer was not properly initialized";
-             sciPrintText((unsigned char *) str, strlen(str));
+             sciPrintf("%s\n",str);
         }
             
 
diff --git a/tempCodeRunnerFile.c b/tempCodeRunnerFile.c
deleted file mode 100644
index dfa038f..0000000
--- a/tempCodeRunnerFile.c
+++ /dev/null
@@ -1,10 +0,0 @@
-
-#include <stdio.h>
-
-int main () {
-
-    int a = 23;
-    int *var; 
-    var = &a;
-    printf("%x", a);
-}
\ No newline at end of file

```

caiicar293 committed:
integrated comments

dgobalak commented on drivers/source/serial_io.c:
```diff
@@ -19,7 +19,7 @@
 
```
Looks like you have some build errors to fix. Btw, you can build your program locally by following the instructions [here](https://www.notion.so/uworbital/Firmware-Onboarding-Challenge-b208ef6d62cd436a8e738dc99f6d2c4d?pvs=4#4cd6898892814e87adbad8122e3004dd).

caiicar293 committed:
onboard challenge 3

dgobalak commented on drivers/source/serial_io.c:
```diff
@@ -64,9 +66,22 @@ obc_error_code_t sciPrintText(unsigned char *text, uint32_t length) {
+        sciSendBytes(text, length);
+        xSemaphoreGive(mutex);
+    } else {
+        return OBC_ERR_CODE_MUTEX_CREATION_FAILED;
```
There's a more appropriate error code for this. If we failed to take the mutex, that means we timed out. It doesn't necessarily mean the mutex wasn't created.

dgobalak commented on onboarding/source/controller.c:
```diff
@@ -97,16 +116,29 @@ static void controllerTask(void * pvParameters) {
     if (lightServiceStatus != OBC_ERR_CODE_SUCCESS) {
         /* USER CODE BEGIN */
-        // Deal with error when initializing light service task and/or queue
+        char str[] = "Error: Light service was not properly initialized";
```
No need to declare an array for the string. You can just pass `"Error: Light service was not properly initialized"` into `sciPrintf`

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -17,12 +26,23 @@
 // Define light service queue config here
-
+#define LIGHT_QUEUE_LENGTH 10
+#define LIGHT_QUEUE_SIZE sizeof(light_event_t)
```
It'd be clearer to name this `LIGHT_QUEUE_ITEM_SIZE` since it's just the size of a single item in the queue

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -17,12 +26,23 @@
+
+static QueueHandle_t lightserviceQueue;
+static StaticQueue_t xlightStaticQueue;
+uint8_t uclightQueueStorageArea[LIGHT_QUEUE_LENGTH * TEXT_SIZE];
```
Use LIGHT_QUEUE_SIZE instead of TEXT_SIZE

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +53,83 @@ static void lightServiceTask(void * pvParameters);
+        lightserviceQueue = xQueueCreateStatic(LIGHT_QUEUE_LENGTH, LIGHT_QUEUE_SIZE,uclightQueueStorageArea, &xlightStaticQueue);
+
+    if (lightserviceQueue == NULL) {
+        return OBC_ERR_CODE_INVALID_ARG;
```
Return an error code saying we failed to create the queue.

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +53,83 @@ static void lightServiceTask(void * pvParameters);
+    }
 
-    /* USER CODE END */
+    if (lightserviceQueue == NULL) {
```
Indentation here is off

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +53,83 @@ static void lightServiceTask(void * pvParameters);
+/* USER CODE END */
+}
+
+uint32_t getLightSensorData(void)
```
`adcData.value` is a `uint16_t` if I remember correctly

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +53,83 @@ static void lightServiceTask(void * pvParameters);
+                uint16_t lightdata = getLightSensorData();
+                    snprintf(str, 32, "The ambient light is %u", lightdata);
+                    /** - Transmit the Conversion data to PC using SCI*/
+                    sciPrintf("%u\n", str);
```
`sciPrintf` basically works the same as `stdio`'s `printf` function. Read more about it here: https://cplusplus.com/reference/cstdio/printf/

No need to use `snprintf` here. Also, no need for the `str` array

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +53,83 @@ static void lightServiceTask(void * pvParameters);
+            return OBC_ERR_CODE_QUEUE_FULL;
+        }
+    }
+    return OBC_ERR_CODE_SUCCESS;
```
If `lightserviceQueue` is NULL, we shouldn't return success. Create a new error code called `OBC_ERR_CODE_INVALID_STATE` and return that if `lightserviceQueue` is NULL

dgobalak reviewed (changes requested):
Looks good! Just a few comments.

caiicar293 committed:
Signed-off-by: Timothy Manuel <timothymanuel295@gmail.com>

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +53,82 @@ static void lightServiceTask(void * pvParameters);
 obc_error_code_t sendToLightServiceQueue(light_event_t *event) {
     /* USER CODE BEGIN */
-    // Send the event to the queue. Return error code if event was not sent successfully.
-    
```
Check if `event` is NULL

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +53,82 @@ static void lightServiceTask(void * pvParameters);
+    if(lightserviceQueue != NULL){
+        if(xQueueSend(lightserviceQueue, event, portMAX_DELAY == pdTRUE)){
+            return OBC_ERR_CODE_SUCCESS;
+        }
```
Return `OBC_ERR_CODE_QUEUE_FULL` if `xQueueSend` fails

dgobalak commented on onboarding/source/controller.c:
```diff
@@ -97,16 +116,27 @@ static void controllerTask(void * pvParameters) {
     if (lightServiceStatus != OBC_ERR_CODE_SUCCESS) {
         /* USER CODE BEGIN */
-        // Deal with error when initializing light service task and/or queue
+        sciPrintf("%s\n","Error: Light service was not properly initialized");
```
Instead of this, you can just write `sciPrintf("Error: Light service was not properly initialized\n");`

dgobalak replied:
This applies to the other times you print error messages as well. Although, this isn't a big deal so don't worry about it for this challenge.

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -17,12 +26,23 @@
+
+static QueueHandle_t lightserviceQueue;
+static StaticQueue_t xlightStaticQueue;
+uint8_t uclightQueueStorageArea[LIGHT_QUEUE_LENGTH * LIGHT_QUEUE_SIZE];
```
The size of this array should be able to store all the items in the queue. Each item in the queue is of type `light_event_t` and there are a maximum of `LIGHT_QUEUE_LENGTH` items. So the size of this array needs to be at least `LIGHT_QUEUE_LENGTH * LIGHT_QUEUE_ITEM_SIZE `

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -17,12 +26,23 @@
-
+#define LIGHT_QUEUE_LENGTH 10
+#define LIGHT_QUEUE_ITEM_SIZE sizeof(light_event_t)
+#define LIGHT_QUEUE_SIZE sizeof( uint64_t )
```
No need for this macro

dgobalak commented on onboarding/source/amb_light_service.c:
```diff
@@ -33,21 +53,82 @@ static void lightServiceTask(void * pvParameters);
+/* USER CODE END */
+
+
+uint16_t getLightSensorData(void)
```
Btw nice job making a helper function for this. This makes it much easier to debug/test code.

caiicar293 committed:
fixed bugs

dgobalak reviewed (commented):
Nice! Looks good to me. I'll close this PR now
